File: src/java/generator/src/main/java/com/uuidgenerator/lib/sonyflake/Sonyflake.java
package com.uuidgenerator.lib.sonyflake;

import com.uuidgenerator.lib.IdGenerator;
import com.uuidgenerator.lib.NetworkUtil;

import java.util.concurrent.atomic.AtomicLong;

public class Sonyflake implements IdGenerator {
    private static final long EPOCH = 1767225600000L; // Jan 1, 2026
    private static final long MACHINE_ID_BITS = 16L;
    private static final long SEQUENCE_BITS = 8L;
    private static final long MAX_MACHINE_ID = (1L << MACHINE_ID_BITS) - 1L;
    private static final long MAX_SEQUENCE = (1L << SEQUENCE_BITS) - 1L;
    private static final long MACHINE_ID_SHIFT = SEQUENCE_BITS;
    private static final long TIMESTAMP_SHIFT = SEQUENCE_BITS + MACHINE_ID_BITS;

    private final long machineId;
    private final AtomicLong sequence = new AtomicLong(0L);
    private final AtomicLong lastTimestamp = new AtomicLong(-1L);

    public Sonyflake() {
        this.machineId = NetworkUtil.getNodeIdFromIp() & MAX_MACHINE_ID;
    }

    private long currentTime10ms() {
        return System.currentTimeMillis() / 10L;
    }

    private long waitForNext10ms(long lastTs) {
        long timestamp = currentTime10ms();
        while (timestamp <= lastTs) {
            timestamp = currentTime10ms();
        }
        return timestamp;
    }

    public long nextId() {
        long timestamp = currentTime10ms();
        long lastTs = lastTimestamp.get();

        if (timestamp < lastTs) {
            System.err.println("Clock moved backwards. Refusing to generate id.");
            return 0L;
        }

        if (timestamp == lastTs) {
            long seq = (sequence.incrementAndGet()) & MAX_SEQUENCE;
            if (seq == 0) {
                timestamp = waitForNext10ms(lastTs);
            }
        } else {
            sequence.set(0L);
        }

        lastTimestamp.set(timestamp);

        return ((timestamp - (EPOCH / 10L)) << TIMESTAMP_SHIFT) |
               (machineId << MACHINE_ID_SHIFT) |
               sequence.get();
    }

    @Override
    public String nextIdString() {
        return Long.toUnsignedString(nextId());
    }
}

File: src/node/generator/lib/sonyflake/sonyflake.js
const { getNodeIdFromIp } = require('../network_util');

class Sonyflake {
    constructor() {
        this.EPOCH = 1767225600000n; // Jan 1, 2026
        this.MACHINE_ID_BITS = 16n;
        this.SEQUENCE_BITS = 8n;
        this.MAX_MACHINE_ID = (1n << this.MACHINE_ID_BITS) - 1n;
        this.MAX_SEQUENCE = (1n << this.SEQUENCE_BITS) - 1n;
        this.MACHINE_ID_SHIFT = this.SEQUENCE_BITS;
        this.TIMESTAMP_SHIFT = this.SEQUENCE_BITS + this.MACHINE_ID_BITS;

        this.machineId = BigInt(getNodeIdFromIp()) & this.MAX_MACHINE_ID;
        this.sequence = 0n;
        this.lastTimestamp = -1n;
    }

    currentTime10ms() {
        return BigInt(Math.floor(Date.now() / 10));
    }

    waitForNext10ms(lastTs) {
        let timestamp = this.currentTime10ms();
        while (timestamp <= lastTs) {
            timestamp = this.currentTime10ms();
        }
        return timestamp;
    }

    async nextIdString() {
        let timestamp = this.currentTime10ms();

        if (timestamp < this.lastTimestamp) {
            console.error("Clock moved backwards. Refusing to generate id.");
            return "0";
        }

        if (timestamp === this.lastTimestamp) {
            this.sequence = (this.sequence + 1n) & this.MAX_SEQUENCE;
            if (this.sequence === 0n) {
                timestamp = this.waitForNext10ms(this.lastTimestamp);
            }
        } else {
            this.sequence = 0n;
        }

        this.lastTimestamp = timestamp;

        const id = ((timestamp - (this.EPOCH / 10n)) << this.TIMESTAMP_SHIFT) |
                   (this.machineId << this.MACHINE_ID_SHIFT) |
                   this.sequence;

        return id.toString();
    }
}

module.exports = Sonyflake;
